harness/determined/common/api/_util.py

```diff
@@ -0,0 +1,51 @@
+import enum
+from typing import Callable, Iterator, TypeVar, Union
+
+from determined.common.api import bindings
+
+
+class PageOpts(str, enum.Enum):
+    single = "1"
+    all = "all"
+
+
+# All paginated response types which want read_paginated() to work with.
+# A Paginated union member must have a .pagination attribute.
+# hint: grep 'pagination: .*v1Pagination.*' bindings.py -B 7 | sed -E -n -e 's/class (.*):/\1/p'
+Paginated = Union[
+    bindings.v1GetAgentsResponse,
+    bindings.v1GetCommandsResponse,
+    bindings.v1GetExperimentCheckpointsResponse,
+    bindings.v1GetExperimentTrialsResponse,
+    bindings.v1GetExperimentsResponse,
+    bindings.v1GetJobsResponse,
+    bindings.v1GetModelVersionsResponse,
+    bindings.v1GetModelsResponse,
+    bindings.v1GetNotebooksResponse,
+    bindings.v1GetResourcePoolsResponse,
+    bindings.v1GetShellsResponse,
+    bindings.v1GetTemplatesResponse,
+    bindings.v1GetTensorboardsResponse,
+    bindings.v1GetTrialCheckpointsResponse,
+    bindings.v1GetTrialWorkloadsResponse,
+    bindings.v1GetUsersResponse,
+    bindings.v1GetWorkspaceProjectsResponse,
+    bindings.v1GetWorkspacesResponse,
+]
+
+T = TypeVar("T", bound=Paginated)
+
+
+def read_paginated(
+    get_with_offset: Callable[[int], T],
+    offset: int = 0,
+    pages: PageOpts = PageOpts.all,
+) -> Iterator[T]:
+    while True:
+        resp = get_with_offset(offset)
+        pagination = resp.pagination
+        assert pagination
+        yield resp
+        if pagination.endIndex == pagination.total or pages == PageOpts.single:
+            break
+        offset = resp.endIndex + 1
```

Comment:
should be         `pagination.endIndex + 1`

Reply:
yeah, I hit this in manual testing.

Also, experimentally I find that it should be `pagination.endIndex`, without the `+ 1`.  Do you know why that is?  Otherwise I seem to miss one entry after every page (which is obvious when `--limit=1`).

Reply:
I don't know enough about where endIndex comes from, but if you received one result you would want to OFFSET 1 and not OFFSET (1+1), so maybe it makes sense

Reply:
@stoksc would you also double-check this?  I'm a bit scared to land it without confirmation of this detail.

Reply:
confirmed with @stoksc, `endIndex` has the same "one after the end" mechanics as python slice indexing.